File: tools/collection-indexer/src/njupt_search_indexer/sitegraph_binary_index.py
from __future__ import annotations

import json
from typing import Any


PACKED_IMPACT_MAGIC_V1 = b"SGIXB001"
PACKED_IMPACT_MAGIC_V2 = b"SGIXB002"
PACKED_IMPACT_MAGIC = PACKED_IMPACT_MAGIC_V2


def encode_varint(value: int) -> bytes:
    if value < 0:
        raise ValueError("varint value must be non-negative")
    chunks = bytearray()
    current = value
    while current >= 0x80:
        chunks.append((current & 0x7F) | 0x80)
        current >>= 7
    chunks.append(current)
    return bytes(chunks)


def decode_varint(data: bytes, offset: int) -> tuple[int, int]:
    shift = 0
    value = 0
    current_offset = offset
    while current_offset < len(data):
        byte = data[current_offset]
        current_offset += 1
        value |= (byte & 0x7F) << shift
        if (byte & 0x80) == 0:
            return value, current_offset
        shift += 7
        if shift > 63:
            raise ValueError("varint is too large")
    raise ValueError("truncated varint")


def impact_index_metadata(payload: dict[str, Any]) -> dict[str, Any]:
    return {
        key: value
        for key, value in payload.items()
        if key != "terms"
    }


def _pack_term_fields(fields: dict[str, Any], term: str) -> bytes:
    out = bytearray()
    if not isinstance(fields, dict):
        raise ValueError(f"impact term fields must be an object: {term}")
    out.extend(encode_varint(len(fields)))
    for field, doc_ids in sorted(fields.items()):
        field_code = str(field)
        if len(field_code.encode("ascii")) != 1:
            raise ValueError(f"field code must be one ASCII byte: {field_code}")
        sorted_ids = sorted(int(item) for item in doc_ids)
        out.extend(field_code.encode("ascii"))
        out.extend(encode_varint(len(sorted_ids)))
        previous = 0
        for index, doc_id in enumerate(sorted_ids):
            delta = doc_id if index == 0 else doc_id - previous
            if delta < 0:
                raise ValueError("doc ids must be sorted")
            out.extend(encode_varint(delta))
            previous = doc_id
    return bytes(out)


def pack_impact_index(payload: dict[str, Any]) -> bytes:
    metadata = json.dumps(impact_index_metadata(payload), ensure_ascii=False, separators=(",", ":")).encode("utf-8")
    terms = payload.get("terms")
    if not isinstance(terms, dict):
        raise ValueError("impact index payload missing terms")

    out = bytearray(PACKED_IMPACT_MAGIC)
    out.extend(len(metadata).to_bytes(4, "little"))
    out.extend(metadata)
    packed_terms: list[tuple[str, bytes, bytes]] = []
    for term, fields in sorted(terms.items()):
        term_bytes = str(term).encode("utf-8")
        packed_terms.append((str(term), term_bytes, _pack_term_fields(fields, str(term))))

    out.extend(encode_varint(len(packed_terms)))
    for _term, term_bytes, payload_bytes in packed_terms:
        out.extend(encode_varint(len(term_bytes)))
        out.extend(term_bytes)
        out.extend(encode_varint(len(payload_bytes)))
    for _term, _term_bytes, payload_bytes in packed_terms:
        out.extend(payload_bytes)
    return bytes(out)


def _unpack_term_fields(data: bytes, offset: int, end: int, *, collect: bool = True) -> tuple[dict[str, list[int]], int]:
    field_count, offset = decode_varint(data, offset)
    fields: dict[str, list[int]] = {}
    for _ in range(field_count):
        if offset >= end:
            raise ValueError("packed impact field code is truncated")
        field = chr(data[offset])
        offset += 1
        doc_count, offset = decode_varint(data, offset)
        doc_ids: list[int] = []
        previous = 0
        for index in range(doc_count):
            delta, offset = decode_varint(data, offset)
            if offset > end:
                raise ValueError("packed impact doc ids are truncated")
            doc_id = delta if index == 0 else previous + delta
            if collect:
                doc_ids.append(doc_id)
            previous = doc_id
        if collect:
            fields[field] = doc_ids
    if offset != end:
        raise ValueError("packed impact term payload has trailing bytes")
    return fields, offset


def _selected_term_bytes(selected_terms: set[str] | None) -> set[bytes] | None:
    if selected_terms is None:
        return None
    return {term.encode("utf-8") for term in selected_terms}


def _unpack_v1_terms(data: bytes, offset: int, selected_terms: set[str] | None = None) -> tuple[dict[str, dict[str, list[int]]], int]:
    term_count, offset = decode_varint(data, offset)
    selected_bytes = _selected_term_bytes(selected_terms)
    terms: dict[str, dict[str, list[int]]] = {}
    for _ in range(term_count):
        term_length, offset = decode_varint(data, offset)
        term_end = offset + term_length
        if term_end > len(data):
            raise ValueError("packed impact term is truncated")
        term_bytes = data[offset:term_end]
        offset = term_end
        collect = selected_bytes is None or term_bytes in selected_bytes
        term = term_bytes.decode("utf-8") if collect else ""
        field_count, offset = decode_varint(data, offset)
        fields: dict[str, list[int]] = {}
        for _ in range(field_count):
            if offset >= len(data):
                raise ValueError("packed impact field code is truncated")
            field = chr(data[offset])
            offset += 1
            doc_count, offset = decode_varint(data, offset)
            doc_ids: list[int] = []
            previous = 0
            for index in range(doc_count):
                delta, offset = decode_varint(data, offset)
                doc_id = delta if index == 0 else previous + delta
                if collect:
                    doc_ids.append(doc_id)
                previous = doc_id
            if collect:
                fields[field] = doc_ids
        if collect:
            terms[term] = fields
    return terms, offset


def _unpack_v2_terms(data: bytes, offset: int, selected_terms: set[str] | None = None) -> tuple[dict[str, dict[str, list[int]]], int]:
    term_count, offset = decode_varint(data, offset)
    selected_bytes = _selected_term_bytes(selected_terms)
    directory: list[tuple[str, int, bool]] = []
    payload_length_total = 0
    for _ in range(term_count):
        term_length, offset = decode_varint(data, offset)
        term_end = offset + term_length
        if term_end > len(data):
            raise ValueError("packed impact term is truncated")
        term_bytes = data[offset:term_end]
        offset = term_end
        payload_length, offset = decode_varint(data, offset)
        collect = selected_bytes is None or term_bytes in selected_bytes
        term = term_bytes.decode("utf-8") if collect else ""
        directory.append((term, payload_length, collect))
        payload_length_total += payload_length
    payload_start = offset
    if payload_start + payload_length_total != len(data):
        raise ValueError("packed impact payload directory length mismatch")

    terms: dict[str, dict[str, list[int]]] = {}
    for term, payload_length, collect in directory:
        payload_end = offset + payload_length
        if payload_end > len(data):
            raise ValueError("packed impact term payload is truncated")
        if collect:
            fields, _ = _unpack_term_fields(data, offset, payload_end, collect=True)
            terms[term] = fields
        offset = payload_end
    return terms, offset


def _unpack_impact_index(data: bytes, selected_terms: set[str] | None = None) -> dict[str, Any]:
    if data.startswith(PACKED_IMPACT_MAGIC_V2):
        magic = PACKED_IMPACT_MAGIC_V2
    elif data.startswith(PACKED_IMPACT_MAGIC_V1):
        magic = PACKED_IMPACT_MAGIC_V1
    else:
        raise ValueError("packed impact index has invalid magic header")
    offset = len(magic)
    if offset + 4 > len(data):
        raise ValueError("packed impact index is missing metadata length")
    metadata_length = int.from_bytes(data[offset: offset + 4], "little")
    offset += 4
    metadata_end = offset + metadata_length
    if metadata_end > len(data):
        raise ValueError("packed impact index metadata is truncated")
    payload = json.loads(data[offset:metadata_end])
    offset = metadata_end
    terms, offset = _unpack_v2_terms(data, offset, selected_terms) if magic == PACKED_IMPACT_MAGIC_V2 else _unpack_v1_terms(data, offset, selected_terms)
    if offset != len(data):
        raise ValueError("packed impact index has trailing bytes")
    payload["terms"] = terms
    return payload


def unpack_impact_index(data: bytes) -> dict[str, Any]:
    return _unpack_impact_index(data)


def unpack_impact_terms(data: bytes, terms: list[str] | set[str]) -> dict[str, Any]:
    return _unpack_impact_index(data, set(terms))

File: packages/search-core/src/sitegraphBinaryIndex.ts
import type { SitegraphImpactIndex, SitegraphLocalBodyIndex } from '@njupt-search/contracts';

const MAGIC_V1 = 'SGIXB001';
const MAGIC_V2 = 'SGIXB002';
const MAGIC_LENGTH = 8;
const decoder = new TextDecoder();

class BinaryCursor {
    private offset = 0;

    constructor(private readonly data: Uint8Array, private readonly source: string) {}

    readMagic(): string {
        const actual = decoder.decode(this.readBytes(MAGIC_LENGTH));
        if (actual !== MAGIC_V1 && actual !== MAGIC_V2) {
            throw new Error(`Validation failed for ${this.source}: invalid packed impact index header`);
        }
        return actual;
    }

    readUint32(): number {
        const b0 = this.data[this.offset];
        const b1 = this.data[this.offset + 1];
        const b2 = this.data[this.offset + 2];
        const b3 = this.data[this.offset + 3];
        if (b0 === undefined || b1 === undefined || b2 === undefined || b3 === undefined) {
            throw new Error(`Validation failed for ${this.source}: truncated uint32`);
        }
        this.offset += 4;
        const value = b0 | (b1 << 8) | (b2 << 16) | (b3 << 24);
        return value >>> 0;
    }

    readVarint(): number {
        let shift = 0;
        let value = 0;
        while (this.offset < this.data.length) {
            const byte = this.data[this.offset];
            if (byte === undefined) {
                throw new Error(`Validation failed for ${this.source}: truncated varint`);
            }
            this.offset += 1;
            value += (byte & 0x7f) * 2 ** shift;
            if ((byte & 0x80) === 0) return value;
            shift += 7;
            if (shift > 53) {
                throw new Error(`Validation failed for ${this.source}: varint exceeds safe integer range`);
            }
        }
        throw new Error(`Validation failed for ${this.source}: truncated varint`);
    }

    readByte(): number {
        const byte = this.data[this.offset];
        if (byte === undefined) {
            throw new Error(`Validation failed for ${this.source}: truncated byte`);
        }
        this.offset += 1;
        return byte;
    }

    readBytes(length: number): Uint8Array {
        if (this.offset + length > this.data.length) {
            throw new Error(`Validation failed for ${this.source}: truncated bytes`);
        }
        const value = this.data.subarray(this.offset, this.offset + length);
        this.offset += length;
        return value;
    }

    done(): boolean {
        return this.offset === this.data.length;
    }

    position(): number {
        return this.offset;
    }
}

const readTermFields = (
    cursor: BinaryCursor,
    collect: boolean
): Record<string, number[]> | null => {
    const fields: Record<string, number[]> = {};
    const fieldCount = cursor.readVarint();
    for (let fieldIndex = 0; fieldIndex < fieldCount; fieldIndex += 1) {
        const field = String.fromCharCode(cursor.readByte());
        const docCount = cursor.readVarint();
        const docIds: number[] = collect ? [] : [];
        let previous = 0;
        for (let docOffset = 0; docOffset < docCount; docOffset += 1) {
            const delta = cursor.readVarint();
            const docId = docOffset === 0 ? delta : previous + delta;
            if (collect) docIds.push(docId);
            previous = docId;
        }
        if (collect) fields[field] = docIds;
    }
    return collect ? fields : null;
};

const readV1Terms = (
    cursor: BinaryCursor,
    source: string,
    selectedTerms?: Set<string>
): SitegraphImpactIndex['terms'] => {
    const termCount = cursor.readVarint();
    const terms: SitegraphImpactIndex['terms'] = {};
    for (let termIndex = 0; termIndex < termCount; termIndex += 1) {
        const term = decoder.decode(cursor.readBytes(cursor.readVarint()));
        const collect = !selectedTerms || selectedTerms.has(term);
        const fields = readTermFields(cursor, collect);
        if (fields) terms[term] = fields;
    }
    if (!cursor.done()) {
        throw new Error(`Validation failed for ${source}: trailing bytes in packed impact index`);
    }
    return terms;
};

const readV2Terms = (
    data: Uint8Array,
    cursor: BinaryCursor,
    source: string,
    selectedTerms?: Set<string>
): SitegraphImpactIndex['terms'] => {
    const termCount = cursor.readVarint();
    const directory: Array<{ length: number; term: string }> = [];
    let payloadLengthTotal = 0;
    for (let termIndex = 0; termIndex < termCount; termIndex += 1) {
        const term = decoder.decode(cursor.readBytes(cursor.readVarint()));
        const length = cursor.readVarint();
        directory.push({ length, term });
        payloadLengthTotal += length;
    }
    const payloadStart = cursor.position();
    if (payloadStart + payloadLengthTotal !== data.length) {
        throw new Error(`Validation failed for ${source}: packed impact payload directory length mismatch`);
    }
    const terms: SitegraphImpactIndex['terms'] = {};
    let offset = payloadStart;
    for (const entry of directory) {
        const collect = !selectedTerms || selectedTerms.has(entry.term);
        if (collect) {
            const payloadCursor = new BinaryCursor(data.subarray(offset, offset + entry.length), `${source}:${entry.term}`);
            const fields = readTermFields(payloadCursor, true);
            if (fields) terms[entry.term] = fields;
        }
        offset += entry.length;
    }
    return terms;
};

const readPackedImpactIndex = <TIndex extends SitegraphImpactIndex = SitegraphImpactIndex>(
    buffer: ArrayBuffer,
    source: string,
    selectedTerms?: Set<string>
): TIndex => {
    const data = new Uint8Array(buffer);
    const cursor = new BinaryCursor(data, source);
    const magic = cursor.readMagic();
    const metadataLength = cursor.readUint32();
    const metadata = JSON.parse(decoder.decode(cursor.readBytes(metadataLength))) as Omit<TIndex, 'terms'>;
    const terms = magic === MAGIC_V2
        ? readV2Terms(data, cursor, source, selectedTerms)
        : readV1Terms(cursor, source, selectedTerms);
    return {
        ...metadata,
        terms,
    } as TIndex;
};

export const decodePackedImpactIndex = <TIndex extends SitegraphImpactIndex = SitegraphImpactIndex>(
    buffer: ArrayBuffer,
    source = 'packed impact index'
): TIndex => readPackedImpactIndex<TIndex>(buffer, source);

export const decodePackedImpactIndexTerms = <TIndex extends SitegraphImpactIndex = SitegraphImpactIndex>(
    buffer: ArrayBuffer,
    terms: string[],
    source = 'packed impact index'
): TIndex => readPackedImpactIndex<TIndex>(buffer, source, new Set(terms));

export const decodePackedLocalBodyIndex = (buffer: ArrayBuffer, source = 'packed local body index'): SitegraphLocalBodyIndex => {
    return decodePackedImpactIndex<SitegraphLocalBodyIndex>(buffer, source);
};

export const decodePackedLocalBodyIndexTerms = (
    buffer: ArrayBuffer,
    terms: string[],
    source = 'packed local body index'
): SitegraphLocalBodyIndex => {
    return decodePackedImpactIndexTerms<SitegraphLocalBodyIndex>(buffer, terms, source);
};
